File: Grind75/Week-1/20.valid-parentheses.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution
{
public:
    bool isValid(string s)
    {
        stack<int> st;
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{')
            {
                st.push(s[i]);
            }
            else
            {
                if (st.empty())
                {
                    return false;
                }
                if (st.top() == '(' && s[i] == ')')
                {
                    st.pop();
                }
                else if (st.top() == '{' && s[i] == '}')
                {
                    st.pop();
                }
                else if (st.top() == '[' && s[i] == ']')
                {
                    st.pop();
                }
                else
                {
                    return false;
                }
            }
        }
        return st.empty();
    }
};
// @lc code=end

File: Java/Grind75/Week 1/20.valid-parentheses.java
import java.util.*;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (Character character : s.toCharArray()) {
            if (character == '{' || character == '[' || character == '(')
                stack.push(character);
            else if (stack.empty())
                return false;
            else if ((character.equals(')') && stack.peek().equals('('))
                    || (character.equals(']') && stack.peek().equals('['))
                    || (character.equals('}') && stack.peek().equals('{')))
                stack.pop();
            else
                return false;
        }
        return stack.empty();
    }
}
// @lc code=end
